openedx/core/djangoapps/content/block_structure/models.py

```diff
@@ -134,14 +193,17 @@ def update_or_create(cls, serialized_data, data_usage_key, **kwargs):
         for the given data_usage_key in the kwargs,
         uploading serialized_data as the content data.
         """
-        bs_model, created = cls.objects.update_or_create(defaults=kwargs, data_usage_key=data_usage_key)
-        bs_model.data.save('', ContentFile(serialized_data))
-        log.info(
-            'BlockStructure: %s in store; %r, size: %d',
-            'Created' if created else 'Updated',
-            unicode(bs_model),
-            len(serialized_data),
-        )
+        # Use an atomic transaction so the model isn't updated
+        # unless the file is successfully persisted.
+        with transaction.atomic():
```

Comment:
**Note to reviewers:** Previously, I had incorrectly assumed that the model wouldn't be saved unless the file was successfully persisted - since the platform's default is to have request-wide atomic transactions.  However, this doesn't always seem to be the case.  And, there may be callers that override this default.  So, I've explicitly added the transaction here, to be safe.